### DISCUSSION TITLE
Edit section fails when fenced code block contains comments

### DISCUSSION THREAD

**Discussion by chrisv2** [django-wiki/django-wiki#1245] [Bug] [answered]
The following markup will break section editing (due to the `#` in the fenced code block):

````markdown
# Section 1
Section 1 Lorem ipsum dolor sit amet

```python
# hello world
print("hello world")
```

# Section 2
Section 2 Lorem ipsum dolor sit amet
````

# Expected result
When clicking the "edit" link next to "Section 2", the editor should show the following markup:

```markdown
# Section 2
Section 2 Lorem ipsum dolor sit amet
```

# Actual result

The editor shows parts of the fenced code block:

````markdown
# hello world
print("hello world")
```

# Section 2
Section 2 Lorem ipsum dolor sit amet
````

# Notes

Actually it is surprisingly hard to create a proper fix for that, as [it is not possible to get source line numbers from the markdown parser](https://github.com/Python-Markdown/markdown/issues/1235). Currently, the editsection plugin parses the source itself by using regexes, but that doesn't account for fenced code blocks, and it's probably quite difficult to do this with regex alone. IMHO we could do one of the following:

1. loop over markdown source line by line, detecting headers and start/end of fenced code blocks (and possibly other blocks like HTML)
2. create a block level extension and annotate each `h[1-6]` with the source line, then find them in the markup

I'm interested in hacking at this, but I'm not sure which is the best approach. Any thoughts?

**Comment by benjaoming** [comment]
Thanks for investigating this!!

I moved it to its own issue. @oscarmcm it's really weird but the "Create issue from discussion" doesn't @-mention the original author and doesn't create a reference in the Discussion. It also doesn't close the discussion or create a redirect. I thought it had these functions, I'm confused :)

Anyways, here's the new issue:
https://github.com/django-wiki/django-wiki/issues/1246

**Reply by oscarmcm** [reply]
Hi @benjaoming, sorry for the late reply I was out for vacations... I'll take a look on this Discussion-Issue workflow.

**Reply by benjaoming** [reply]
If you want to quickly try it then open a Discussion and convert it to an issue, perhaps the easiest way to see what I mean :)

(or in case I was wrong)

**Reply by benjaoming** [reply]
It's really cool that Discussions have threads -- issue comments do not :) That's a big feature!

**Comment by chrisv2** [comment]
To make more clear what I mean with option 2.), it would be something like this:

```python
class AnnotateHeadingsProcessor(BlockProcessor):
    def __init__(self, parser):
        super().__init__(parser)
        self.cur_block, self.last_block = None, None

    def test(self, parent, block):
        # save the last block and current block
        self.last_block = self.cur_block
        self.cur_block = block
        return True

    def run(self, parent, blocks):
        last_child = self.lastChild(parent)
        if last_child is not None and last_child.tag.lower() in ('h1', 'h2', 'h3', 'h4', 'h5', 'h6'):
            # the other parsers added a hX block, so we must find out what has been consumed
            # => build difference between last_block and cur_block
            if self.last_block.endswith(self.cur_block):
                source_line = self.last_block[:len(self.last_block)-len(self.cur_block)].strip()
                last_child.attrib['data-source'] = source_line

        # we don't consume any blocks by ourselves
        return False
```

We can also compute the source line number from that, and use it later in tree processor

**Comment by benjaoming** [comment]
Very interesting solution!

> create a block level extension and annotate each h[1-6] with the source line, then find them in the markup

Is this an "internal" extension that's only used for parsing the Markdown when we want to edit a particular section?

**Reply by chrisv2** [reply]
Well I intended to have it as part of `EditSectionExtension`, unfortunately this approach doesn't work all the time, due to the implementation of `SetextHeaderProcessor` and `HashHeaderProcessor` the following markup

```markdown
Title 1
=======
## Title 2
```
will cause a recursive call to my block processor, leaving me unable to properly keep track of the consumed source lines.

I suppose we could use [CommonMark's AST](https://waylonwalker.com/python-markdown-ast-paragraph/) to get the source positions of  all headings, but that would mean having a second parser, which doesn't seem very elegant to me.

**Reply by chrisv2** [reply]
Found a better solution (https://github.com/django-wiki/django-wiki/pull/1247)

**Comment by oscarmcm** [comment, marked as answer]
Fixed via #1247
